ipa-core/src/report/hybrid_info.rs

```diff
@@ -2,43 +2,95 @@ use crate::report::{hybrid::NonAsciiStringError, KeyIdentifier};
 
 const DOMAIN: &str = "private-attribution";
 
-#[derive(Debug)]
-pub struct HybridImpressionInfo<'a> {
+#[derive(Clone, Debug)]
+pub struct HybridImpressionInfo {
     pub key_id: KeyIdentifier,
-    pub helper_origin: &'a str,
+    pub helper_origin: &'static str,
 }
 
-#[allow(dead_code)]
+impl HybridImpressionInfo {
+    /// Creates a new instance.
+    ///
+    /// ## Errors
+    /// if helper or site origin is not a valid ASCII string.
+    pub fn new(
+        key_id: KeyIdentifier,
+        helper_origin: &'static str,
+    ) -> Result<Self, NonAsciiStringError> {
+        // If the types of errors returned from this function change, then the validation in
+        // `EncryptedReport::from_bytes` may need to change as well.
+        if !helper_origin.is_ascii() {
+            return Err(helper_origin.into());
+        }
+
+        Ok(Self {
+            key_id,
+            helper_origin,
+        })
+    }
+
+    // Converts this instance into an owned byte slice that can further be used to create HPKE
+    // sender or receiver context.
+    pub(super) fn to_bytes(&self) -> Box<[u8]> {
+        let info_len = DOMAIN.len()
+            + self.helper_origin.len()
+            + 2 // delimiters(?)
+            + std::mem::size_of_val(&self.key_id);
+        let mut r = Vec::with_capacity(info_len);
+
+        r.extend_from_slice(DOMAIN.as_bytes());
+        r.push(0);
+        r.extend_from_slice(self.helper_origin.as_bytes());
+        r.push(0);
+
+        r.push(self.key_id);
+
+        debug_assert_eq!(r.len(), info_len, "HPKE Info length estimation is incorrect and leads to extra allocation or wasted memory");
+
+        r.into_boxed_slice()
+    }
+}
+
+#[derive(Clone, Debug)]
 pub struct HybridConversionInfo<'a> {
     pub key_id: KeyIdentifier,
-    pub helper_origin: &'a str,
-    pub converion_site_domain: &'a str,
+    pub helper_origin: &'static str,
+    pub conversion_site_domain: &'a str,
```

Comment:
There can be many conversion sites in publisher queries - we never properly implemented it and I believe just hardcoded both helper origin and conversion site. I think this is fine for now, just wanted to explain it